sql/row: use Put instead of CPut when updating value of secondary index

michae2 opened the pull request:
**sql/row: use Put instead of CPut when updating value of secondary index**

When an UPDATE statement changes the value but not the key of a secondary index (e.g. an update to the stored columns of a secondary index) we need to write a new version of the secondary index KV with the new value.

We were using a CPutAllowingIfNotExists to do this, which verified that if the KV existed, the expected value was the value before update. But there's no need for this verification. We have other mechanisms to detect a write-write conflict with any other transaction that could have changed the value concurrently. We can simply use a Put to overwrite the previous value.

This also matches what we do for the primary index when the PK doesn't change.

Epic: None

Release note: None

---

**sql: change CPutAllowingIfNotExists with nil expValue to CPut**

CPutAllowingIfNotExists with empty expValue is equivalent to CPut with empty expValue, so change a few spots to use regular CPut. This almost gets rid of CPutAllowingIfNotExists entirely, but there is at least one spot in backfill (introduced in #138707) that needs to allow for both a non-empty expValue and non-existence of the KV.

Also drop "(expecting does not exist)" from CPut tracing, as CPut with empty expValue is now overwhelmingly the most common use of CPut. And this matches the tracing in #138707.

Epic: None

Release note: None

michae2 committed:
sql/row: use Put instead of CPut when updating value of secondary index

When an UPDATE statement changes the value but not the key of a
secondary index (e.g. an update to the stored columns of a secondary
index) we need to write a new version of the secondary index KV with the
new value.

We were using a CPutAllowingIfNotExists to do this, which verified that
if the KV existed, the expected value was the value before update. But
there's no need for this verification. We have other mechanisms to
detect a write-write conflict with any other transaction that could have
changed the value concurrently. We can simply use a Put to overwrite the
previous value.

This also matches what we do for the primary index when the PK doesn't
change.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137805)
<!-- Reviewable:end -->

michae2 committed:
sql: change CPutAllowingIfNotExists with nil expValue to CPut

CPutAllowingIfNotExists with empty expValue is equivalent to CPut with
empty expValue, so change a few spots to use regular CPut. This almost
gets rid of CPutAllowingIfNotExists entirely, but there is at least one
spot in backfill (introduced in #138707) that needs to allow for both a
non-empty expValue and non-existence of the KV.

Also drop "(expecting does not exist)" from CPut tracing, as CPut with
empty expValue is now overwhelmingly the most common use of CPut. And
this matches the tracing in #138707.

Epic: None

Release note: None

michae2 commented:
I think this is RFAL.

yuzefovich reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137805#-:-OGIQtPL5BRC6YaJtfLd:be8rotp)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/secondary_index_column_families` line 231 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137805#-OGIQ80t3_6PkZ47v2Fb:-OGIQ80u-XfUWsQ4hKjm:bzgrdpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed7e0bce4956036c55e8132c03a933a1bcedd276/pkg/sql/opt/exec/execbuilder/testdata/secondary_index_column_families#L231)):*
> ```
> 
> # Ensure that updates only touch the changed column families.
> query T kvtrace(CPut,prefix=/Table/107/2/)
> ```

nit: let's update the directive to match the `Put`.


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/sql/row/updater.go:
```diff
@@ -428,13 +428,17 @@ func (ru *Updater) UpdateRow(
 						if traceKV {
 							k := keys.PrettyPrint(ru.Helper.secIndexValDirs[i], newEntry.Key)
 							v := newEntry.Value.PrettyPrint()
-							if expValue != nil {
-								log.VEventf(ctx, 2, "CPut %s -> %v (replacing %v, if exists)", k, v, oldEntry.Value.PrettyPrint())
+							if sameKey {
+								log.VEventf(ctx, 2, "Put %s -> %v", k, v)
 							} else {
-								log.VEventf(ctx, 2, "CPut %s -> %v (expecting does not exist)", k, v)
+								log.VEventf(ctx, 2, "CPut %s -> %v", k, v)
 							}
```
Not for this PR, but if we ever do a cleanup pass of this function, it would be nice to re-organize things so the logging doesn't take up so much attention.  Also of interest is that we log ForcePut's differently.

stevendanna reviewed (approved):
Nice!

michae2 commented:
TFTRs!

bors r=yuzefovich,stevendanna

michae2 commented:
Trying again...

bors r=yuzefovich,stevendanna
